mlir/python/PyCompilerDriver.cpp

```diff
@@ -0,0 +1,108 @@
+// limitations under the License.
+
+#include <iostream>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "mlir/IR/BuiltinTypes.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include "Catalyst/Driver/CompilerDriver.h"
+
+namespace py = pybind11;
+
+std::vector<Pipeline> parseCompilerSpec(const py::list &pipelines)
+{
+    std::vector<Pipeline> out;
+    for (py::handle obj : pipelines) {
+        py::tuple t = obj.cast<py::tuple>();
+        auto i = t.begin();
+        auto py_name = i;
+        i++;
+        auto py_passes = i;
+        i++;
+        assert(i == t.end());
+        std::string name = py_name->attr("__str__")().cast<std::string>();
+        Pipeline::PassList passes;
+        std::transform(py_passes->begin(), py_passes->end(), std::back_inserter(passes),
+                       [](py::handle p) { return p.attr("__str__")().cast<std::string>(); });
+        out.push_back(Pipeline({name, passes}));
+    }
+    return out;
+}
+
+PYBIND11_MODULE(compiler_driver, m)
+{
+    //===--------------------------------------------------------------------===//
+    // Catalyst Compiler Driver
+    //===--------------------------------------------------------------------===//
+    py::class_<FunctionAttributes> funcattrs_class(m, "FunctionAttributes");
+    funcattrs_class.def(py::init<>())
+        .def("get_function_name",
+             [](const FunctionAttributes &fa) -> std::string { return fa.functionName; })
+        .def("get_return_type",
+             [](const FunctionAttributes &fa) -> std::string { return fa.returnType; });
+
+    py::class_<CompilerOutput> compout_class(m, "CompilerOutput");
+    compout_class.def(py::init<>())
+        .def("get_pipeline_output",
+             [](const CompilerOutput &co, const std::string &name) -> std::optional<std::string> {
+                 auto res = co.pipelineOutputs.find(name);
+                 return res != co.pipelineOutputs.end() ? res->second
+                                                        : std::optional<std::string>();
+             })
+        .def("get_output_ir", [](const CompilerOutput &co) -> std::string { return co.outIR; })
+        .def("get_object_filename",
+             [](const CompilerOutput &co) -> std::string { return co.objectFilename; })
+        .def("get_function_attributes",
+             [](const CompilerOutput &co) -> FunctionAttributes { return co.inferredAttributes; })
+        .def("get_diagnostic_messages",
+             [](const CompilerOutput &co) -> std::string { return co.diagnosticMessages; });
+
+    m.def(
+        "run_compiler_driver",
+        [](const char *source, const char *workspace, const char *moduleName, bool keepIntermediate,
+           bool verbose, py::list pipelines, py::list llvmPipelines,
+           bool lower_to_llvm) -> CompilerOutput * {
+            FunctionAttributes inferredAttributes;
+            std::string errors;
+
+            CompilerOutput *output = new CompilerOutput();
```

Comment:
I missed this in my first pass... let's wrap it in a std::unique_ptr

Reply:
https://github.com/PennyLaneAI/catalyst/pull/216/commits/44d6bde89a08f19e6d86070cfcc341f32a83b1cc